atc/engine/build_step_delegate.go

```diff
@@ -373,11 +373,29 @@ func (delegate *buildStepDelegate) ConstructAcrossSubsteps(templateBytes []byte,
 		if err := yaml.Unmarshal(interpolatedBytes, &subPlan); err != nil {
 			return nil, fmt.Errorf("invalid template bytes: %w", err)
 		}
+
+		// Maps from the original subplan ID generated by the planner to the
+		// translated ID unique to the substep iteration.
+		mappedSubplanIDs := map[atc.PlanID]atc.PlanID{}
 		planIDCounter := 0
 		subPlan.Each(func(p *atc.Plan) {
-			p.ID = atc.PlanID(fmt.Sprintf("%s/%d/%d", delegate.planID, i, planIDCounter))
+			mappedID := atc.PlanID(fmt.Sprintf("%s/%d/%d", delegate.planID, i, planIDCounter))
+			mappedSubplanIDs[p.ID] = mappedID
+			p.ID = mappedID
 			planIDCounter++
 		})
+
+		subPlan.Each(func(p *atc.Plan) {
```

Comment:
we have two loops of subPlan, is it possible to combine them?

Reply:
Combining them probably *would* still work, but it would introduce some coupling, since we'd need to depend on the order that `atc.Plan.Each` calls the function in. e.g. if `atc.Plan.Each` starting doing a pre-order/post-order traversal, this would break things.

When we do a first pass for re-mapping the IDs, we know that `mappedSubplanIDs` has processed *all* of the subplans. This makes it safe to rely on in the second pass for the `VersionFrom` mapping, regardless of how `atc.Plan.Each` traverses the plans.

That said, maybe that's fine - it's nice to know what type of traversal a function will use, so maybe that should just be a part of the contract of the method. But by the same token, iterating twice also isn't that big a deal, and won't really be a performance burden